File: app/domain/relations/compound_connect.py
"""填詞連接詞複合（!{連接}!／~{連接}~）— ADR-0053：詞庫 ∪ 合成 + syn/ant 互斥。"""

from __future__ import annotations

from typing import Dict, Literal

from sqlalchemy.orm import Session

from app.domain.relations.compound_syn import narrow_compound_syn_literals
from app.models.word import Word
from app.services._generated.fillword_connectives import FILLWORD_CONNECTIVES

# ponytail: guide／essay 常見填詞連接詞三字詞 — ensure 入庫（CONTEXT § 連接詞複合查詢）
CONNECTIVE_LITERAL_SEEDS: dict[str, tuple[str, ...]] = {
    "與": ("生與死", "天與地", "男與女", "父與子"),
}

# Lexicon flank tiers 0–2; synthetic always ranks after
TIER_CONNECTIVE_SYNTH = 3
# Grill: avoid freezing Portable search on first !與! / ~與~
CONNECTIVE_SYNTH_CAP = 500


def _flank_tiers_from_two_char(two_char_tiers: Dict[str, int]) -> Dict[tuple[str, str], int]:
    out: Dict[tuple[str, str], int] = {}
    for w, tier in two_char_tiers.items():
        if len(w) != 2:
            continue
        a, b = w[0], w[1]
        for pair in ((a, b), (b, a)):
            prev = out.get(pair)
            out[pair] = tier if prev is None else min(prev, tier)
    return out


def exclusive_two_char_tiers(
    primary: Dict[str, int],
    opposite: Dict[str, int],
    kind: Literal["syn", "ant"],
) -> Dict[str, int]:
    """Strict mutual exclusion with ant-wins: syn drops ant pairs; ant keeps full primary."""
    if kind == "ant":
        return dict(primary)
    return {w: t for w, t in primary.items() if w not in opposite}


def _three_char_literals(db: Session) -> set[str]:
    rows = db.query(Word.char).filter(Word.length == 3).distinct().all()
    return {row[0] for row in rows if row[0] and len(row[0]) == 3}


def search_connective_compound(
    db: Session,
    *,
    compound_kind: Literal["syn", "ant"],
    connective: str,
    rhyme_char: str | None = None,
) -> Dict[str, int]:
    """!與!／~與~：詞庫三字 ∩ exclusive flank ∪ 合成缺席 A連B（cap）。"""
    if connective not in FILLWORD_CONNECTIVES:
        return {}

    from app.domain.lexicon.port import default_word_inject_port
    from app.domain.relations.compound_ant import search_compound_ant
    from app.domain.relations.compound_syn import search_compound_syn

    inject = default_word_inject_port()
    for literal in CONNECTIVE_LITERAL_SEEDS.get(connective, ()):
        inject.ensure_word_rows(db, literal)

    syn_tiers = search_compound_syn(db)
    ant_tiers = search_compound_ant(db)
    if compound_kind == "ant":
        exclusive = exclusive_two_char_tiers(ant_tiers, syn_tiers, "ant")
    else:
        exclusive = exclusive_two_char_tiers(syn_tiers, ant_tiers, "syn")

    flank_tiers = _flank_tiers_from_two_char(exclusive)
    if not flank_tiers:
        return {}

    tiers: Dict[str, int] = {}
    for w in _three_char_literals(db):
        if len(w) != 3 or w[1] != connective:
            continue
        tier = flank_tiers.get((w[0], w[2]))
        if tier is None:
            continue
        tiers[w] = tier

    synth_n = 0
    for (a, b), _flank in flank_tiers.items():
        if synth_n >= CONNECTIVE_SYNTH_CAP:
            break
        compound = f"{a}{connective}{b}"
        if compound in tiers:
            continue
        rows = inject.ensure_word_rows(db, compound)
        if rows:
            tiers[compound] = TIER_CONNECTIVE_SYNTH
            synth_n += 1

    if not rhyme_char:
        return tiers
    allowed = narrow_compound_syn_literals(
        frozenset(tiers.keys()), width=3, rhyme_char=rhyme_char, db=db
    )
    return {ch: tiers[ch] for ch in allowed if ch in tiers}


__all__ = [
    "CONNECTIVE_SYNTH_CAP",
    "FILLWORD_CONNECTIVES",
    "TIER_CONNECTIVE_SYNTH",
    "exclusive_two_char_tiers",
    "search_connective_compound",
]

File: client/src/db/compound-connect.ts
/**
 * 連接詞複合（~與~／!與!）— ADR-0053：詞庫 ∪ 合成 + syn/ant 互斥。
 * Port of app/domain/relations/compound_connect.py
 */
import type { Database } from './sqljs.ts';
import { queryRows } from './database-backend.ts';
import { ensureConnectiveCompoundRows, ensureComposedWordRow } from './db-patch.ts';
import { FILLWORD_CONNECTIVES_SET as FILLWORD_CONNECTIVES } from './_generated/fillword-connectives.ts';

export type ConnectiveCompoundKind = 'syn' | 'ant';

/** Lexicon hits keep flank tier 0–2; synthetic always ranks after. */
export const TIER_CONNECTIVE_SYNTH = 3;

/** Grill: cap synth so first !與! / ~與~ does not freeze main thread. */
export const CONNECTIVE_SYNTH_CAP = 500;
const SYNTH_YIELD_EVERY = 25;

type TierMap = Map<string, number>;

function yieldToMain(): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, 0));
}

function flankTiersFromTwoChar(twoCharTiers: TierMap): Map<string, number> {
  const out = new Map<string, number>();
  for (const [w, tier] of twoCharTiers) {
    if (w.length !== 2) continue;
    for (const pair of [`${w[0]}\t${w[1]}`, `${w[1]}\t${w[0]}`]) {
      const prev = out.get(pair);
      out.set(pair, prev === undefined ? tier : Math.min(prev, tier));
    }
  }
  return out;
}

/**
 * Strict mutual exclusion with ant-wins:
 * - syn path: drop any pair also on ant (反義唔入 ~)
 * - ant path: keep full ant primary (近義污染唔踢走 !! 對)
 */
export function exclusiveTwoCharTiers(
  primary: TierMap,
  opposite: TierMap,
  kind: ConnectiveCompoundKind,
): TierMap {
  if (kind === 'ant') {
    return new Map(primary);
  }
  const out = new Map<string, number>();
  for (const [w, tier] of primary) {
    if (!opposite.has(w)) {
      out.set(w, tier);
    }
  }
  return out;
}

async function loadThreeCharLiterals(db: Database): Promise<Set<string>> {
  const rows = await queryRows(
    db,
    `
    SELECT DISTINCT char FROM words
    WHERE length = 3 OR ((length IS NULL OR length = 0) AND length(char) = 3)
  `,
  );
  const out = new Set<string>();
  for (const row of rows) {
    const ch = String(row.char ?? '');
    if (ch.length === 3) out.add(ch);
  }
  return out;
}

/** Cache key → tiers; cleared via resetConnectiveCompoundCache. */
const connectiveTierCache = new Map<string, TierMap>();

export function resetConnectiveCompoundCache(): void {
  connectiveTierCache.clear();
}

/**
 * ~{連}~／!{連}!：詞庫三字 ∩ exclusive flank，再合成缺席 A連B。
 */
export async function searchConnectiveCompoundTiers(
  db: Database,
  opts: {
    compoundKind: ConnectiveCompoundKind;
    connective: string;
    synTiers: TierMap;
    antTiers: TierMap;
  },
): Promise<TierMap> {
  const { compoundKind, connective, synTiers, antTiers } = opts;
  if (!connective || !FILLWORD_CONNECTIVES.has(connective)) {
    return new Map();
  }

  const cacheKey = `${compoundKind}\0${connective}\0${synTiers.size}\0${antTiers.size}`;
  const cached = connectiveTierCache.get(cacheKey);
  if (cached) {
    return cached;
  }

  await ensureConnectiveCompoundRows(db);

  const primary = compoundKind === 'ant' ? antTiers : synTiers;
  const opposite = compoundKind === 'ant' ? synTiers : antTiers;
  const exclusive = exclusiveTwoCharTiers(primary, opposite, compoundKind);
  const flankTiers = flankTiersFromTwoChar(exclusive);
  if (!flankTiers.size) {
    const empty = new Map<string, number>();
    connectiveTierCache.set(cacheKey, empty);
    return empty;
  }

  const tiers = new Map<string, number>();
  for (const w of await loadThreeCharLiterals(db)) {
    if (w[1] !== connective) continue;
    const tier = flankTiers.get(`${w[0]}\t${w[2]}`);
    if (tier !== undefined) {
      tiers.set(w, tier);
    }
  }

  // Lexicon hits first (already in `tiers`); synth up to CAP with main-thread yields
  const pending: string[] = [];
  for (const [pair] of flankTiers) {
    if (pending.length >= CONNECTIVE_SYNTH_CAP) break;
    const tab = pair.indexOf('\t');
    if (tab < 0) continue;
    const a = pair.slice(0, tab);
    const b = pair.slice(tab + 1);
    if (!a || !b) continue;
    const compound = `${a}${connective}${b}`;
    if (!tiers.has(compound)) {
      pending.push(compound);
    }
  }
  let n = 0;
  for (const compound of pending) {
    const ok = await ensureComposedWordRow(db, compound);
    if (ok) {
      tiers.set(compound, TIER_CONNECTIVE_SYNTH);
    }
    n += 1;
    if (n % SYNTH_YIELD_EVERY === 0) {
      await yieldToMain();
    }
  }

  connectiveTierCache.set(cacheKey, tiers);
  return tiers;
}
